src/main/java/webserver/Controller.java

```diff
@@ -21,20 +34,87 @@ public ResponseEntity service(final RequestEntity request) {
+        }
+        final byte[] file = fileService.serveFile(request.getPath());
+        return ResponseEntity.of(request.getPath(), file);
+    }
+
+    private ResponseEntity getUserList(final RequestEntity request) throws IOException {
+        if (!isLogin(request)) {
+            return ResponseEntity.redirectResponseEntity("/user/login.html");
+        }
+        final Collection<User> users = userService.findAll();
+        final byte[] file = fileService.render(request.getPath(), Map.of("users", users));
+        return ResponseEntity.of(request.getPath(), file);
+    }
+
     private ResponseEntity createUser(final RequestEntity request) {
         if (request.isGet()) {
-            userService.createUser(new CreateUserDto(request.getHeader().getQueryParameter()));
+            validateCreateUser(request.getQueryParameters());
+            userService.createUser(CreateUserDto.of(request.getQueryParameters()));
+            return ResponseEntity.redirectResponseEntity("/index.html");
+        }
+        if (request.isPost()) {
+            validateCreateUser(request.getBody().get());
+            userService.createUser(CreateUserDto.of(request.getBody().get()));
+            return ResponseEntity.redirectResponseEntity("/index.html");
         }
-        userService.createUser(new CreateUserDto(request.getBody().get()));
-        return ResponseEntity.redirectResponseEntity("/index.html");
+
+        throw new IllegalArgumentException("존재하지 않는 요청입니다.");
+    }
+
+    private void validateCreateUser(final Map<String, String> values) {
+        if (values.get("userId") == null ||
+                values.get("password") == null ||
+                values.get("name") == null ||
+                values.get("email") == null) {
+            throw new IllegalArgumentException("입력값이 올바르지 않습니다.");
+        }
+    }
+
+    private ResponseEntity login(final RequestEntity request) {
+        if (request.isPost()) {
+            try {
+                final User user = userService.login(new LoginDto(request.getBody().get("userId"), request.getBody().get("password")));
+                final HttpSession httpSession = new HttpSession();
+                httpSession.setAttribute("user", user);
+                SessionManager.addSession(httpSession);
+
+                ResponseEntity responseEntity = ResponseEntity.redirectResponseEntity("/index.html");
+                responseEntity.setCookie(new Cookies(Map.of(COOKIE_SESSION_KEY, httpSession.getId()), "/"));
+                return responseEntity;
+            } catch (IllegalArgumentException e) {
+                return ResponseEntity.redirectResponseEntity("/user/login_failed.html");
+            }
+        }
+
+        throw new IllegalArgumentException("존재하지 않는 요청입니다.");
+    }
+
+    private boolean isLogin(final RequestEntity request) {
+        return SessionManager.getSession(request.getCookies().getCookie(COOKIE_SESSION_KEY)) != null;
```

Comment:
`2.템플릿 엔진 활용하기` 에서 요구하고있는 로그인을 판별하는 로직과 조금 다른 것 같아요. 🤔

또한 (위에서 코메트한 것처럼) `COOKIE_SESSION_KEY`를 `Cookies` 객체가 가지고  `cookies.getSession()`을 해줄 수 있겠네요. 
그렇다면 request 객체에게 `getSession()` 메시지를 보내면 cookies 객체의 getSession() 메서드를 포워딩해주기만하면 바로 세션을 가져올 수 있겠네요. 

```suggestion
        return SessionManager.getSession(request.getSession()) != null;
```